lib/SILOptimizer/Mandatory/AddressLowering.cpp

```diff
@@ -282,312 +537,659 @@ class OpaqueValueVisitor {
+    SILValue value, ArrayRef<SILValue> incomingValues, bool intoPhi) {
+  // Def-projections take precedence.
+  assert(!getProjectedDefOperand(value) && !getReusedStorageOperand(value));
+
+  for (Operand *use : value->getUses()) {
+    // Get the user's value, whose storage we will project into.
+    SILValue userValue = getProjectedUseValue(use);
+    if (!userValue)
+      continue;
+
+    assert(!getProjectedDefOperand(userValue)
+           && "storage cannot project in two directions.");
+
+    // Recurse through all storage projections to find the uniquely allocated
+    // storage. Enum storage cannot be reused across multiple subobjects because
+    // it must be initialized via a single init_enum_data_addr instruction.
+    //
+    // TODO: fix the memory verifier to consider the actual store instructions
+    // to initialize an enum rather than the init_enum_data_addr to reuse enum
+    // storage across multiple subobjects within the payload.
+    auto *baseStorage = pass.valueStorageMap.getBaseStorage(
+        userValue, /*allowInitEnum*/ !intoPhi);
+    if (!baseStorage)
+      continue;
+
+    if (auto *stackInst =
+            dyn_cast<AllocStackInst>(baseStorage->storageAddress)) {
+      if (!checkStorageDominates(stackInst, incomingValues))
+        continue;
+    } else
+      assert(isa<SILFunctionArgument>(baseStorage->storageAddress));
+
+    LLVM_DEBUG(llvm::dbgs() << "  PROJECT "; value->dump();
+               llvm::dbgs() << "  into use "; use->getUser()->dump());
+
+    pass.valueStorageMap.recordComposingUseProjection(use, userValue);
+    return true;
   }
+  return false;
+}
 
-  if (value->hasOneUse()) {
-    // TODO: Handle block arguments.
-    // TODO: Handle subobjects with a single composition, and other non-mutating
-    // uses such as @in arguments.
-    if (auto *def = dyn_cast<SingleValueInstruction>(value)) {
-      Operand *useOper = *value->use_begin();
-      if (canProjectFrom(def, useOper->getUser())) {
-        storage.setComposedOperand(useOper);
-        return;
-      }
+bool OpaqueStorageAllocation::
+checkStorageDominates(AllocStackInst *allocInst,
+                      ArrayRef<SILValue> incomingValues) {
+
+  for (SILValue incomingValue : incomingValues) {
+    if (auto *defInst = incomingValue->getDefiningInstruction()) {
+      if (!pass.domInfo->properlyDominates(allocInst, defInst))
+        return false;
+      continue;
+    }
+    // Handle both phis and terminator results.
+    auto *bbArg = cast<SILPhiArgument>(incomingValue);
+    // The storage block must strictly dominate the phi.
+    if (!pass.domInfo->properlyDominates(
+          allocInst->getParent(), bbArg->getParent())) {
+      return false;
+    }
+  }
+  return true;
+}
+
+void OpaqueStorageAllocation::allocatePhi(PhiValue phi) {
+  // Coalesces phi operand storage with the phi storage. The algorithm processes
+  // all incoming values at once, so it is is run when visiting the block
+  // argument.
```

Comment:
Yes, the PhiStorageOptimizer was not well documented at the top level. I'm pasting the file comment here because I can't refer to commits after rebasing
```
/// PhiStorageOptimizer implements an analysis used by AddressLowering
/// to reuse storage across block arguments.
///
/// In OSSA, phi operands can often be coalesced because they are
/// consuming--they end the lifetime of their operand. This optimization may
/// fail to coalesce an operand for two major reasons:
///
/// 1. This phi operand is already coalesced with other storage, possibly of a
/// different type:
///
///     %field = struct_extract %struct : $Struct<T>, #field
///     br bb(%field : $T)
///
///   bb(%phi : @owned $T):
///     ...
///
/// 2. This phi operand interferes with another coalesced phi operand.
///
/// Only one of the call results below, either %get0 or %get1, can be coalesced
/// with %phi. The %phi will itself be coalesced with this function's indirect
/// @out argument.
///
///   sil [ossa] @function : $@convention(thin) <T> () -> @out T {
///   bb0:
///     %get0 = apply %get<T>() : $@convention(thin) <τ_0_0>() -> @out τ_0_0
///     %get1 = apply %get<T>() : $@convention(thin) <τ_0_0>() -> @out τ_0_0
///     cond_br undef, bb2, bb1
///
///   bb1:
///     destroy_value %get0 : $T
///     br bb3(%get1 : $T)
///
///   bb2:
///     destroy_value %get1 : $T
///     br bb3(%get0 : $T)
///
///   bb3(%phi : @owned $T):
///     return %phi : $T
///
/// TODO: Liveness is currently recorded at the block level. This could be
/// extended to handle operand with nonoverlapping liveness in the same
/// block. In this case, %get0 and %get1 could both be coalesced with a bit of
/// extra book-keeping:
///
///   bb0:
///     %get0 = apply %get<T>() : $@convention(thin) <τ_0_0>() -> @out τ_0_0
///
///   bb1:
///     destroy_value %get0 : $T
///     %get1 = apply %get<T>() : $@convention(thin) <τ_0_0>() -> @out τ_0_0
///     br bb3(%get1 : $T)
///
///   bb2:
///     br bb3(%get0 : $T)
///
///   bb3(%phi : @owned $T):
```